### ISSUE TITLE
add combinator to easily chain ZIO[R, Option[E], T]

### ISSUE THREAD

**Issue by TimPigden** [zio/zio#3338]
One way of giving an idiomatic partial function in zio is potentially to have a function return 
```
ZIO[R, Option[E], T]
```
If the operation matches, then return succeed(t), if it fails fail(Some(e)) and if there's no match then fail(None)
This can then be chained to another function which does the same thing.
A zio combinator would inspect the result of the first function and call the second only for E(None) - you'd probably need some sort of NotFound at the end of the chain to go with it. So you could chain to 
```
effect1(someArgs) orNotMatched effect1(someArgs) orNoMatch
```

**Comment by adamgfraser** [comment]
@TimPigden Thanks for raising. Looking at this now.

**Comment by adamgfraser** [comment]
How about?

```scala
  /**
   * Returns an effect that will produce the value of this effect, unless it
   * fails with the `None` value, in which case it will produce the value of
   * the specified effect.
   */
  final def orElseOptional[R1 <: R, E1, A1 >: A](
    that: => ZIO[R1, Option[E1], A1]
  )(implicit ev: E <:< Option[E1]): ZIO[R1, Option[E1], A1] =
    catchAll(ev(_).fold(that)(e => ZIO.fail(Some(e))))
```

**Comment by adamgfraser** [comment]
Resolved by #3339.
